File: ISSUE-60/SOLUTION-5/kty.py
#!/usr/bin/env python3
#-*- coding:utf-8 -*-
#
# 2014-07-06 by kantal59 
# License: LGPL
# OLIMEX WPC#60: PTC linearization
#			KTY81/110 is non linear resistor which increases it’s value with temperature. 
#			You can read the datasheet here: http://www.nxp.com/documents/data_sheet/KTY81_SER.pdf
# 			Make code which input KTY81/110 real number of resistance in ohms and output the temperature of the sensor.
#
#
# 			Resources: 
#			 	http://www.nxp.com/documents/data_sheet/KTY81_SER.pdf
#			 	http://www.nxp.com/documents/other/SC17_GENERAL_TEMP_1996_3.pdf
# 	

import sys

def mktbl_kty81_110():
	''' Creating table of (resistance,temperature) values.
		Formula used:	RT = Rref*( 1 + A*( T – Tref ) + B*( T – Tref )^2 – C*( T – Ti )^D )	where  C==0 if T < Ti 
		Sensor current: 1mA
		(It is easy to create a similar procedure for other KTY devices if the parameters are given.)
	'''
	Tref= 25
	Rref= 1000
	Ti= 100
	A= 7.874e-3
	B= 1.874e-5
	C= 3.42e-8 
	D= 3.7
	Trange= range( -56, 151)
	prec= 2	#  one unit is divided into 'prec' equal part

	tbl=[]	

	for t in Trange:
		for k in range(prec):
			T= t + k/prec
			if T >= Ti:
				tbl.append( ( Rref*( 1+A*(T-Tref)+B*(T-Tref)*(T-Tref)-C*pow(T-Ti, D) ), T ) )
			else:
				tbl.append( ( Rref*( 1+A*(T-Tref)+B*(T-Tref)*(T-Tref) ), T ) )

	return( tbl)				



def get_temperature( ohm, tbl):
	''' It returns the temperature value from the table "tbl" corresponding to resistance "ohm". 
		"tbl" has (resistance,temperature) values.
		It can be used for other KTY devices with different tables, too.
	'''
	if ohm < tbl[0][0]  or  ohm > tbl[-1][0] :	return( None)		# invalid input resistance value
	
	v= (0,tbl[0])	# (tbl index, (resistance,temperature) )
	
	for i,(r,t) in enumerate(tbl):
		if r > ohm:	break;				
		v= ( i,(r,t) )
		
	k,(r1,t1) = v;
	if ohm == r1:	# the given resistance is listed in the table
		return( t1)

	r2,t2 = tbl[k+1];
	return( t1+ (t2-t1)*(ohm-r1)/(r2-r1) )



TBL_kty81_110= mktbl_kty81_110()


# ------------ TEST ------------
if __name__ == '__main__':

	R=[ 100,490, 814.85, 1000,1002,1010,1012,1020,1028,1029,1040, 1300, 1700, 1900, 2200, 2300 ] # test data
	if len(sys.argv) < 2:
		print(" Usage: $ ./kty.py  resistance1 resistance2 ...\n  where the values are in Ohms")
	else:
		R=[]
		for i in range( 1, len(sys.argv)):
			R.append( float(sys.argv[i]))		


	print("KTY81_110:")
	#print("KTY81_250:")
	for r in R:
		t= get_temperature( r,TBL_kty81_110 )
		# t= get_temperature( r,TBL_kty81_250 )
		if t:	t= "{0:.2f} Celsius".format(t)
		print( " {0} Ohms\t{1}".format( r, t)) 	

			

File: ISSUE-60/SOLUTION-11/kty.c
/*
 2014-07-06 by kantal59 
 License: LGPL
 OLIMEX WPC#60: PTC linearization
			KTY81/110 is non linear resistor which increases it’s value with temperature. 
			You can read the datasheet here: http://www.nxp.com/documents/data_sheet/KTY81_SER.pdf
 			Make code which input KTY81/110 real number of resistance in ohms and output the temperature of the sensor.

 			Resources: see my Python solution kty.py!
 			
 			Compile: gcc -o kty kty.c
*/


#include <stdio.h>
#include <stdlib.h>
#include "kty.h"


float get_temperature( float fohm,	const rtdata_t  *pdata)
{

	unsigned int ohm,i,index,r1,r2;
	int t1,t2;
	size_t size= pdata->len;
	const restemp_t* ptbl= pdata->ptbl;
	
	ohm= fohm*100;
	if( ohm < ptbl[0].res ||  ohm > ptbl[ size-1].res )		return( INVTEMPR);		// invalid input resistance value yields invalid temperature
	
	for( i=0;	i< size;	i++){
	
		if( ptbl[i].res > ohm ) break;
		index= i;
	}

	if( ohm == ptbl[index].res  )	return( (float)ptbl[index].temp / 100 );
	
	r1= ptbl[index].res;	t1= ptbl[index].temp;
	r2= ptbl[index+1].res;	t2= ptbl[index+1].temp;
	
	return( ( (float)t1 + (float)((t2-t1)*(ohm-r1))/(r2-r1)  ) / 100 );

}


int main( int argc, char *argv[]){

	
	if( argc<2 ){ 
					puts(" Usage: $ ./kty.py  resistance1 resistance2 ...\n  where the values are in Ohms");
					return(1);
	}	
	
	int i;			
	for( i=1; i< argc; i++){
		
		float res= strtof( argv[i],NULL);	puts( argv[i]);
		float temp= get_temperature( res, &data_kty81_110);
		if( temp == INVTEMPR )		printf( " %.2f Ohms\tNA\n", res );
		else						printf( " %.2f Ohms\t%.2f\n", res,temp );
			
	}
		
	return(0);
}

